runtime-modules/project-token/src/lib.rs

```diff
@@ -0,0 +1,479 @@
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn transfer(
+            origin,
+            token_id: T::TokenId,
+            outputs: TransfersOf<T>,
+        ) -> DispatchResult {
+            let src = ensure_signed(origin)?;
+
+            // Currency transfer preconditions
+            Self::ensure_can_transfer(token_id, &src, &outputs)?;
+
+            // == MUTATION SAFE ==
+
+            Self::do_transfer(token_id, &src, &outputs);
+
+            Self::deposit_event(RawEvent::TokenAmountTransferred(
+                token_id,
+                src,
+                outputs.into(),
+            ));
+            Ok(())
+        }
+
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn dust_account(origin, token_id: T::TokenId, account_id: T::AccountId) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let token_info = Self::ensure_token_exists(token_id)?;
+            let account_to_remove_info = Self::ensure_account_data_exists(token_id, &account_id)?;
+
+            Self::ensure_user_can_dust_account(
+                &token_info.transfer_policy,
+                &sender,
+                &account_id,
+                &account_to_remove_info,
+            )?;
+
+            // == MUTATION SAFE ==
+
+            AccountInfoByTokenAndAccount::<T>::remove(token_id, &account_id);
+
+            let bloat_bond = T::BloatBond::get();
+            let _ = T::ReserveCurrency::deposit_creating(&account_id, bloat_bond);
+
+            Self::deposit_event(RawEvent::AccountDustedBy(token_id, account_id, sender, token_info.transfer_policy));
+
+            Ok(())
+        }
+
+        /// Join whitelist for permissioned case: used to add accounts for token
+        /// Preconditions:
+        /// - 'token_id' must be valid
+        /// - transfer policy is permissionless or transfer policy is permissioned and merkle proof is valid
+        ///
+        /// Postconditions:
+        /// - account added to the list
+        #[weight = 10_000_000] // TODO: adjust weights
+        pub fn join_whitelist(origin, token_id: T::TokenId, proof: MerkleProofOf<T>) -> DispatchResult {
+            let account_id = ensure_signed(origin)?;
+            let token_info = Self::ensure_token_exists(token_id)?;
+
+            ensure!(
+                !AccountInfoByTokenAndAccount::<T>::contains_key(token_id, &account_id),
+                Error::<T>::AccountAlreadyExists,
+            );
+
+            if let TransferPolicyOf::<T>::Permissioned(commit) = token_info.transfer_policy {
+                proof.verify_for_commit::<T,_>(&account_id, commit)
+            } else {
+                Err(Error::<T>::CannotJoinWhitelistInPermissionlessMode.into())
+            }?;
+
+            let bloat_bond = T::BloatBond::get();
+
+            // No project_token or balances state corrupted in case of failure
+            ensure!(
+                T::ReserveCurrency::free_balance(&account_id) >= bloat_bond,
```

Comment:
The implementation for transfer used in our `substrate` is the same as: https://docs.substrate.io/rustdocs/latest/src/pallet_balances/lib.rs.html#1468
I have added a `ensure_can_transfer_reserve` guard in this case

Reply:
In order to avoid transfer errors when using a Treasury account I used the same method like it is done on storage here: https://github.com/Joystream/joystream/pull/3538/files#